### ISSUE TITLE
`iter.Prefetcher` loses __len__ information

### ISSUE THREAD

**Issue by sehoffmann** [pytorch/data#1114]
### 🐛 Describe the bug

If the upstream pipe has implemented `__len__`, doing `.prefetch()` will loose that information.

We should probably add a mixin class for this, since IterDataPipes that don't modify the element count are quite common.

### Versions

latest

**Comment by NivekT** [comment]
We can potentially do something like:
```python
    def __len__(self) -> int:
        if isinstance(self.datapipe, Sized):
            return len(self.datapipe)
        raise TypeError(
            "{} instance doesn't have valid length".format(type(self).__name__)
```

This is the general pattern for some existing DataPipes (e.g. `.map()`)

**Comment by NivekT** [comment]
Fixed by #1120
